### PR TITLE
Feature: add new init_chg method with wavefunctions

### LINKED ISSUE #5077: A new function is requested according to the pw method calculation of the DFT
### Background

It is very important to obtain the initial density of the system from the fixed occupation number when calculating DFT of the fixed occupation number

### Describe the solution you'd like

It is suggested to define an INPUT parameter in input on the basis of pw, and the initial density of the system can be obtained according to the fixed occupation number (ocp=1)

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Notice Possible Changes of Behavior (Reminder only for developers)

_No response_

### Notice any changes of core modules (Reminder only for developers)

_No response_

### Notice Possible Changes of Core Modules (Reminder only for developers)

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### PR DESCRIPTION
**Pull request by Qianruipku**
### Linked Issue
Fix #5077 :
ABACUS can init charge density from wavefunc*.dat and istate.info.
The occupations can be changed by modifying istate.info.

Other:
Fix bug of read_wfc_pw when kpar > 1
Add more precision for istate.info.

### CODE REVIEW

File: source/module_cell/klist.h [hunk 1]
```diff
@@ -116,7 +117,7 @@ class K_Vectors
      * process pools (KPAR), and adding the remainder if the process pool ID (MY_POOL) is less than the remainder.
      * @note The function is declared as inline for efficiency.
      */
-    inline int getik_global(const int& ik) const;
+    static inline int getik_global(const int& ik, const int& nkstot);
```
Comment (mohanchen) [inline source/module_cell/klist.h]: I would suggest add '_' after 'get', the new name will be get_ik_global

File: source/module_cell/klist.h [hunk 2]
```diff
@@ -390,10 +391,10 @@ class K_Vectors
     void print_klists(std::ofstream& fn);
 };
 
-inline int K_Vectors::getik_global(const int& ik) const
+inline int K_Vectors::getik_global(const int& ik, const int& nkstot)
 {
-    int nkp = this->nkstot / GlobalV::KPAR;
-    int rem = this->nkstot % GlobalV::KPAR;
+    int nkp = nkstot / GlobalV::KPAR;
```
Comment (mohanchen) [inline source/module_cell/klist.h]: avoid using GlobalV?

File: source/module_io/read_wfc_to_rho.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        chg.reduce_diff_pools(chg.rho[is]);
+    }
+#endif
+
+    // Since rho is calculated by psi^2, it is not symmetric. We need to rearrange it. 
+    Symmetry_rho srho;
+    for (int is = 0; is < nspin; is++)
+    {
+        srho.begin(is, chg, chg.rhopw, GlobalC::Pgrid, GlobalC::ucell.symm);
```
Comment (mohanchen) [inline source/module_io/read_wfc_to_rho.cpp]: pass ucell into this function?

File: source/module_io/test/read_wfc_to_rho_test.cpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        //     if (GlobalV::MY_RANK == ip)
+        //     {
+        //         std::cout.precision(16);
+        //         std::cout << GlobalV::MY_RANK << " " << chg.rho[0][0] << std::endl;
+        //     }
+        //     MPI_Barrier(MPI_COMM_WORLD);
+        // }
+    }
+
+    delete[] chg.rho;
```
Comment (mohanchen) [inline source/module_io/test/read_wfc_to_rho_test.cpp:260]: try using std::vector?

Comment (Qianruipku) [reply source/module_io/test/read_wfc_to_rho_test.cpp:260]: Rho is a member of Charge class. It is a double** pointer.

**Review by mohanchen — APPROVED** [review]

**Pull request merged**